partridge/gtfs.py

```diff
@@ -135,32 +135,38 @@ def get(self, filename):
         return df
 
     @contextmanager
-    def _io_adapter(self, filename):
+    def read_file_chunks(self, filename):
         """
-        Yield an IO object for the given file
-        from a zip file or folder
+        Yield a pandas DataFrame iterator for the given file.
         """
-        def reader(iowrapper):
-            """
-            Build a chunked DataFrame reader
-            """
+        with self._io_adapter(self.zmap[filename]) as result:
+            iowrapper, encoding = result
             try:
-                return pd.read_csv(
+                yield pd.read_csv(
                     iowrapper, chunksize=10000,
-                    dtype=np.unicode, index_col=False,
+                    dtype=np.unicode, encoding=encoding, index_col=False,
                     low_memory=False, skipinitialspace=True)
             except pd.errors.EmptyDataError:
-                return iter([])
+                yield iter([])
 
+    @contextmanager
+    def _io_adapter(self, fpath):
+        """
+        Yield an IO object and its encoding for the given file path.
+        Reads from a zip file or folder.
+        """
         if self.is_dir:
-            with open(self.zmap[filename], 'rb') as iowrapper:
-                yield reader(iowrapper)
+            with open(fpath, 'rb') as iowrapper:
+                encoding = detect_encoding(iowrapper)
+                iowrapper.seek(0)  # Rewind to the beginning of the file
+                yield iowrapper, encoding
         else:
             with ZipFile(self.path) as zipreader:
```

Comment:
This method opens the file twice. Once to detect the encoding and again to yield it to the caller. Ideally, it would only be opened once and then rewound after detecting the encoding. Unfortunately, I could not figure out how to rewind the file handle for instances of `zipfile.ZipExtFile`.

Reply:
I wouldn't worry about it. I suspect that the encoding detector will not unwind much of the file and that the zip decoding and reading happens in chunks. It won't be so much duplicated work.

The best you could do to avoid it would be to create your own buffered reader with a large enough buffer to rewind back to the beginning. Probably not worth it.